File: yt/spark/spark-over-yt/java-examples/src/main/java/tech/ytsaurus/spyt/example/GroupingExample.java
package tech.ytsaurus.spyt.example;

import org.apache.spark.SparkConf;
import org.apache.spark.sql.*;

import tech.ytsaurus.spyt.SparkAppJava;
import tech.ytsaurus.client.CompoundClient;

public class GroupingExample extends SparkAppJava {
    @Override
    protected void doRun(String[] args, SparkSession spark, CompoundClient yt) {
        Dataset<Row> df = spark.read().format("yt").load("/sys/spark/examples/example_1");
        Dataset<Row> dictDf = spark.read().format("yt").load("/sys/spark/examples/example_dict");

        df
          .join(dictDf, df.col("uuid").equalTo(dictDf.col("uuid")), "left_outer")
          .groupBy("count")
          .agg(functions.max("id").as("max_id"))
          .repartition(1)
          .write().mode(SaveMode.Overwrite).format("yt").save("/sys/spark/examples/example_1_agg");
    }

    public static void main(String[] args) {
        new GroupingExample().run(args);
    }

    @Override
    protected SparkConf getSparkConf() {
        return super.getSparkConf().setAppName("Custom name");
    }
}

File: ytsaurus/yt/spark/spark-over-yt/scala-examples/src/main/scala/tech/ytsaurus/spyt/examples/GroupingExample.scala
package tech.ytsaurus.spyt.examples

import org.apache.spark.SparkConf
import org.apache.spark.sql.functions._
import org.apache.spark.sql.{SaveMode, SparkSession}
import tech.ytsaurus.spyt._
import tech.ytsaurus.spyt.wrapper.table.OptimizeMode

object GroupingExample extends App {
    val conf = new SparkConf().setAppName("Grouping example")
    val spark = SparkSession.builder.config(conf).getOrCreate()

    val df = spark.read.yt("//home/spark/examples/example_1")
    val dictDf = spark.read.yt("//home/spark/examples/example_dict")

    df
        .join(dictDf, Seq("uuid"), "left_outer")
        .groupBy("count")
        .agg(max("id") as "max_id")
        .repartition(1)
        .write
        .mode(SaveMode.Overwrite)
        .optimizeFor(OptimizeMode.Scan)
        .yt("//home/spark/examples/example_1_agg")
}
